<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Writing iSEE extensions • iSEEDemoEuroBioC2024</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Writing iSEE extensions">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">iSEEDemoEuroBioC2024</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/workshop_isee_extension.html">Writing iSEE extensions</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/iSEE/iSEEDemoEuroBioC2024" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Writing iSEE extensions</h1>
                        <h4 data-toc-skip class="author">Kevin
Rue-Albrecht<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>,Federico Marini<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>, Charlotte Soneson<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/iSEE/iSEEDemoEuroBioC2024/blob/devel/vignettes/workshop_isee_extension.Rmd" class="external-link"><code>vignettes/workshop_isee_extension.Rmd</code></a></small>
      <div class="hidden name"><code>workshop_isee_extension.Rmd</code></div>

    </div>

    
    
<p>Last modified: 09 July, 2024.</p>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h3>
<p>This package demo will present a brief introduction to the
functionality of the <em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em> package
and its existing extension packages, before demonstrating the writing of
new functionality suitable for release in additional extension
packages.</p>
</div>
<div class="section level3">
<h3 id="prerequisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h3>
<p>Workshop prerequisites:</p>
<ul>
<li>Familiarity with the <em><a href="https://bioconductor.org/packages/3.20/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>
class.</li>
<li>Familiarity with the <em><a href="https://CRAN.R-project.org/package=shiny" class="external-link">shiny</a></em>
package</li>
</ul>
<p>Relevant background reading:</p>
<ul>
<li>EuroBioC2023 workshop: <a href="https://isee.github.io/iSEEDemoEuroBioC2023/articles/iSEEdemo.html" class="external-link">materials</a>
</li>
<li>EuroBioC2020 workshop: <a href="https://isee.github.io/iSEEWorkshopEuroBioc2020/" class="external-link">materials</a>,
<a href="https://isee.github.io/iSEEWorkshopEuroBioc2020Slides/" class="external-link">slides</a>
</li>
<li>BioC2020 workshop: <a href="https://isee.github.io/iSEEWorkshop2020/" class="external-link">materials</a>, <a href="https://isee.github.io/iSEEWorkshop2020Slides/" class="external-link">slides</a>
</li>
<li>BioC2019 workshop: <a href="https://isee.github.io/iSEEWorkshop2019/" class="external-link">materials</a>, <a href="https://isee.github.io/iSEEWorkshop2019Slides/" class="external-link">slides</a>
</li>
<li>Rue-Albrecht K, Marini F, Soneson C and Lun ATL. iSEE: Interactive
SummarizedExperiment Explorer [version 1; peer review: 3 approved].
F1000Research 2018, 7:741 (<a href="https://doi.org/10.12688/f1000research.14966.1" class="external-link uri">https://doi.org/10.12688/f1000research.14966.1</a>)</li>
<li>Vignette: <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html#anatomy-of-a-summarizedexperiment" class="external-link">SummarizedExperiment
for Coordinating Experimental Assays, Samples, and Regions of
Interest</a>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="participation">Participation<a class="anchor" aria-label="anchor" href="#participation"></a>
</h3>
<p>Students are encouraged to ask questions throughout the package
demo.</p>
<p>Where applicable, instructors will illustrate answers with live-coded
examples.</p>
<p>Alternatively, students are also encouraged to write questions
before, during, and after the workshop using the ‘New issue’ button on
the GitHub repository for this workshop (<a href="https://github.com/iSEE/iSEEDemoEuroBioC2024/issues" class="external-link uri">https://github.com/iSEE/iSEEDemoEuroBioC2024/issues</a>).</p>
<p>Instructors will respond to GitHub issues at the earliest
opportunity, which may be after the end of the conference.</p>
</div>
<div class="section level3">
<h3 id="r-bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r-bioconductor-packages-used"></a>
</h3>
<ul>
<li><em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.20/iSEEde" class="external-link">iSEEde</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.20/iSEEhex" class="external-link">iSEEhex</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.20/iSEEhub" class="external-link">iSEEhub</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.20/iSEEindex" class="external-link">iSEEindex</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.20/iSEEpathways" class="external-link">iSEEpathways</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.20/iSEEu" class="external-link">iSEEu</a></em></li>
</ul>
</div>
<div class="section level3">
<h3 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h3>
<p>An example for a 40-minute workshop:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>iSEE functionality</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Existing iSEE extensions</td>
<td>10m</td>
</tr>
<tr class="odd">
<td>Writing iSEE extensions</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Questions</td>
<td>10m</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h3>
</div>
<div class="section level3">
<h3 id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h3>
<ul>
<li>Describe how to interactively explore omics data using <em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em>.</li>
<li>Identify extension packages adding functionality to the <em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em>
interface.</li>
<li>Understand what is needed to write <em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em>
extensions.</li>
</ul>
</div>
<div class="section level3">
<h3 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ul>
<li>Launch <em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em>
applications to visualise examples data sets.</li>
<li>Configure <em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em>
applications to use functionality from extension packages.</li>
<li>Create and include a new <em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em> panel
in a live application.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="isee-functionality">iSEE functionality<a class="anchor" aria-label="anchor" href="#isee-functionality"></a>
</h2>
<div class="section level3">
<h3 id="overview-1">Overview<a class="anchor" aria-label="anchor" href="#overview-1"></a>
</h3>
<p>In the next 10 minutes, we will:</p>
<ul>
<li>Load a demonstration data set</li>
<li>Launch an iSEE app in its default state</li>
<li>Export an R script representing a modified app state</li>
<li>Re-launch the same iSEE app preconfigured in that modified
state</li>
</ul>
</div>
<div class="section level3">
<h3 id="input-data">Input data<a class="anchor" aria-label="anchor" href="#input-data"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em> was
designed around the <em><a href="https://bioconductor.org/packages/3.20/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>
class, a container widely used throughout the <em>Bioconductor</em>
project.</p>
<p>Briefly, the <em><a href="https://bioconductor.org/packages/3.20/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>
class provides a container keeping matrices of assay data, sample
metadata, and feature metadata synchronised throughout analytical
workflows (e.g., filtering, reordering).</p>
<div class="float">
<img src="img/SummarizedExperiment.svg" alt="SummarizedExperiment (reproduced from the package vignette; https://bioconductor.org/packages/SummarizedExperiment/)."><div class="figcaption">SummarizedExperiment (reproduced from the
package vignette; <a href="https://bioconductor.org/packages/SummarizedExperiment/" class="external-link uri">https://bioconductor.org/packages/SummarizedExperiment/</a>).</div>
</div>
<p><br></p>
<p>By extension, <em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em>
naturally supports classes derived from <em><a href="https://bioconductor.org/packages/3.20/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>.
For instance, the <em><a href="https://bioconductor.org/packages/3.20/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
class adds functionality for storing matrices of reduced dimensions,
also keeping those synchronised with assay data and metadata during
analyses.</p>
<p>In practice, you would generally create a <em><a href="https://bioconductor.org/packages/3.20/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>
(or derived) object from raw data and metadata loaded from files (e.g.,
RNA-seq count matrix produce by a program like <a href="https://doi.org/10.1093/bioinformatics/btt656" class="external-link">featureCounts</a>
and sample metadata from your lab notebook).</p>
<p><svg aria-hidden="true" role="img" viewbox="0 0 512 512" style="height:1em;width:1em;vertical-align:-0.125em;margin-left:auto;margin-right:auto;font-size:inherit;fill:currentColor;overflow:visible;position:relative;"><path d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM216 336h24V272H216c-13.3 0-24-10.7-24-24s10.7-24 24-24h48c13.3 0 24 10.7 24 24v88h8c13.3 0 24 10.7 24 24s-10.7 24-24 24H216c-13.3 0-24-10.7-24-24s10.7-24 24-24zm40-208a32 32 0 1 1 0 64 32 32 0 1 1 0-64z"></path></svg>
You can learn about the creation and handling of
<code>SummarizedExperiment</code> objects in the <a href="#prerequisites">Relevant background reading</a> resources
above.</p>
<p>In this workshop, we will load a publicly available <em><a href="https://bioconductor.org/packages/3.20/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
object to save some time.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/ReprocessedData.html" class="external-link">ReprocessedAllenData</a></span><span class="op">(</span>assays<span class="op">=</span><span class="st">"tophat_counts"</span><span class="op">)</span></span>
<span><span class="va">sce</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 20816 379 </span></span>
<span><span class="co">#&gt; metadata(2): SuppInfo which_qc</span></span>
<span><span class="co">#&gt; assays(1): tophat_counts</span></span>
<span><span class="co">#&gt; rownames(20816): 0610007P14Rik 0610009B22Rik ... Zzef1 Zzz3</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(379): SRR2140028 SRR2140022 ... SRR2139341 SRR2139336</span></span>
<span><span class="co">#&gt; colData names(22): NREADS NALIGNED ... Animal.ID passes_qc_checks_s</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: endogenous</span></span>
<span><span class="co">#&gt; altExpNames(1): ERCC</span></span></code></pre></div>
<p>In the summary view of the object displayed above, we can tell that
it contains:</p>
<ul>
<li>One assay called <code>tophat_counts</code>, a matrix measuring
20,816 genes in 379 cells.</li>
<li>Twenty two columns of cell metadata.</li>
<li>No gene metadata.</li>
</ul>
<p>Also of note:</p>
<ul>
<li>Gene symbols are in use for the row names.</li>
<li>Arbitrary cell names are in use for column names.</li>
<li>Some object-level metadata is present.</li>
<li>An alternative experiment called <code>ERCC</code> is present (we
will not use it).</li>
</ul>
</div>
<div class="section level3">
<h3 id="getting-started-with-isee">Getting started with iSEE<a class="anchor" aria-label="anchor" href="#getting-started-with-isee"></a>
</h3>
<p>One of the strengths of <em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em> is that
it can be used at any point in an analytical workflow. From the moment
raw data or metadata are encapsulated in a <em><a href="https://bioconductor.org/packages/3.20/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>
object, they’re good to go!</p>
<p>Let’s demonstrate this by attaching the <em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em> package
to the R session and calling the <code><a href="https://rdrr.io/pkg/iSEE/man/iSEE.html" class="external-link">iSEE()</a></code> function to launch
an app in the default settings.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/iSEE/iSEE" class="external-link">iSEE</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/iSEE/man/iSEE.html" class="external-link">iSEE</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="img/iSEE_default.png" alt="Screenshot of iSEE app."><div class="figcaption">Screenshot of iSEE app.</div>
</div>
<p><br></p>
<p>In the screenshot above, you will notice that some of the panels seem
to be truncated at the bottom.</p>
<p><em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em>
applications organise panels along a single-page layout. Following
<em><a href="https://CRAN.R-project.org/package=shiny" class="external-link">shiny</a></em>
rules, pages are divided into 12 units horizontally, allowing up to 12
panels on a row before new rows are added to accommodate more panels.
Depending on the size and set-up of your monitors, you may comfortably
fit a considerable number of panels in your own screen estate.</p>
<p>By default, <em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em>
applications adopt a “showcase” mode that populates the application with
one panel of each type compatible with the assay data and metadata
detected in the input object.</p>
<p>As such, in the screenshot above, we see at least:</p>
<ul>
<li>A ‘Row data table’ panel, displaying gene metadata.</li>
<li>A ‘Feature assay plot’ panel, displaying assayed data for a
gene.</li>
<li>A ‘Column data plot’ panel, displaying a column of cell
metadata.</li>
<li>A ‘Sample assay plot’ panel, displaying assayed data for a
sample.</li>
<li>A ‘Column data table’ panel, displaying cell metadata.</li>
<li>A ‘Complex heatmap’ panel, displaying assayed data for a set of
genes.</li>
</ul>
</div>
<div class="section level3">
<h3 id="core-isee-functionality">Core iSEE functionality<a class="anchor" aria-label="anchor" href="#core-isee-functionality"></a>
</h3>
<p>Within <em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em>
applications, users can interactively:</p>
<ul>
<li>View tables of metadata, filtering rows, and selecting a single row
that may be highlighted in other panels.</li>
<li>View plots of assay data and metadata, controlling data displayed
along each axis, as well as visual aspects such as colour, shape, size,
and faceting, using any suitable data or metadata present in the input
object.</li>
<li>Draw selections in plots using the built-in rectangular <em><a href="https://CRAN.R-project.org/package=shiny" class="external-link">shiny</a></em> brush or
the slower but more flexible lasso selection implemented in <em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em>.</li>
<li>Transmit selection between panels, dynamically controlling which
panels receive information from which panels.</li>
<li>Use transmitted selections to highlight or restrict the
corresponding data points in the receiving panel(s).</li>
<li>Rearrange the layout of panels, adding, removing, or reordering
panels at will.</li>
<li>Resize the panels, in width and height.</li>
</ul>
<p>Furthermore, you will also find functionality for:</p>
<ul>
<li>Downloading the panel outputs: in PDF format for plots, and CSV
format for tables.</li>
<li>Exporting an R script that reproduces exactly each panel output,
including brushes and downstream effects related to transmitted
selections.</li>
<li>Exporting an R script that reproduces exactly the current layout of
the application.</li>
<li>Launching custom interactive tours introducing new users to the
various components of the application.</li>
<li>Opening the vignette of the <em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em> package
without leaving the application.</li>
<li>Inspecting the session information.</li>
<li>Learning more about iSEE, including how to cite it.</li>
<li>Inspecting the overall object metadata (converted to JSON
format).</li>
</ul>
<p>To simulate a short analytical workflow, let us run some more code
from the <code><a href="https://rdrr.io/pkg/iSEE/man/iSEE.html" class="external-link">?iSEE</a></code> help page to:</p>
<ul>
<li>Compute a normalised count matrix.</li>
<li>Compute a PCA.</li>
<li>Compute a t-SNE.</li>
<li>Compute some simple gene metadata.</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span>, exprs_values<span class="op">=</span><span class="st">"tophat_counts"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents<span class="op">=</span><span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">ave_count</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"tophat_counts"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">n_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"tophat_counts"</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">sce</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 20816 379 </span></span>
<span><span class="co">#&gt; metadata(2): SuppInfo which_qc</span></span>
<span><span class="co">#&gt; assays(2): tophat_counts logcounts</span></span>
<span><span class="co">#&gt; rownames(20816): 0610007P14Rik 0610009B22Rik ... Zzef1 Zzz3</span></span>
<span><span class="co">#&gt; rowData names(2): ave_count n_cells</span></span>
<span><span class="co">#&gt; colnames(379): SRR2140028 SRR2140022 ... SRR2139341 SRR2139336</span></span>
<span><span class="co">#&gt; colData names(23): NREADS NALIGNED ... passes_qc_checks_s sizeFactor</span></span>
<span><span class="co">#&gt; reducedDimNames(2): PCA TSNE</span></span>
<span><span class="co">#&gt; mainExpName: endogenous</span></span>
<span><span class="co">#&gt; altExpNames(1): ERCC</span></span></code></pre></div>
<p>Then, as pointed out earlier, you are free to launch another <em><a href="https://bioconductor.org/packages/3.20/iSEE" class="external-link">iSEE</a></em> app
again at any point in an analytical workflow.</p>
<p>The <code><a href="https://rdrr.io/pkg/iSEE/man/iSEE.html" class="external-link">iSEE()</a></code> function automatically detects new assay data
and metadata in the updated object, populating the application
components with all the available information, old and new.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/iSEE/man/iSEE.html" class="external-link">iSEE</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="img/iSEE_default_2.png" alt="Screenshot of iSEE app."><div class="figcaption">Screenshot of iSEE app.</div>
</div>
<p><br></p>
<p>Comparing the screenshot above with the earlier one, you will notice
that the first panel is now a ‘Reduced dimension plot’ panel, and that
all other panels have shifted position down in the order of the layout
(top to bottom, left to right).</p>
<p>Indeed, the last time we launched the app, the object did not contain
any dimensionality reduction result. The <code><a href="https://rdrr.io/pkg/iSEE/man/iSEE.html" class="external-link">iSEE()</a></code> function
automatically detected that, and dropped that type of panel from the
application.</p>
<p>To demonstrate the next functionality, let us interactively (within
the app launched above) remove all the panels except for ‘Reduced
dimension plot 1’ and ‘Feature assay plot 1’. Then, use the download
icon in the navigation bar at the top of the app to find and click the
‘Display panel settings’ button. You will be presented with an R script
that looks as follows:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">initial</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">################################################################################</span></span>
<span><span class="co"># Settings for Reduced dimension plot 1</span></span>
<span><span class="co">################################################################################</span></span>
<span></span>
<span><span class="va">initial</span><span class="op">[[</span><span class="st">"ReducedDimensionPlot1"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">new</span><span class="op">(</span><span class="st">"ReducedDimensionPlot"</span>, Type <span class="op">=</span> <span class="st">"PCA"</span>, XAxis <span class="op">=</span> <span class="fl">1L</span>, YAxis <span class="op">=</span> <span class="fl">2L</span>, </span>
<span>    FacetRowByColData <span class="op">=</span> <span class="st">"driver_1_s"</span>, FacetColumnByColData <span class="op">=</span> <span class="st">"driver_1_s"</span>, </span>
<span>    ColorByColumnData <span class="op">=</span> <span class="st">"NREADS"</span>, ColorByFeatureNameAssay <span class="op">=</span> <span class="st">"logcounts"</span>, </span>
<span>    ColorBySampleNameColor <span class="op">=</span> <span class="st">"#FF0000"</span>, ShapeByColumnData <span class="op">=</span> <span class="st">"driver_1_s"</span>, </span>
<span>    SizeByColumnData <span class="op">=</span> <span class="st">"NREADS"</span>, TooltipColumnData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, </span>
<span>    FacetRowBy <span class="op">=</span> <span class="st">"None"</span>, FacetColumnBy <span class="op">=</span> <span class="st">"None"</span>, ColorBy <span class="op">=</span> <span class="st">"None"</span>, </span>
<span>    ColorByDefaultColor <span class="op">=</span> <span class="st">"#000000"</span>, ColorByFeatureName <span class="op">=</span> <span class="st">"0610007P14Rik"</span>, </span>
<span>    ColorByFeatureSource <span class="op">=</span> <span class="st">"---"</span>, ColorByFeatureDynamicSource <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    ColorBySampleName <span class="op">=</span> <span class="st">"SRR2140028"</span>, ColorBySampleSource <span class="op">=</span> <span class="st">"---"</span>, </span>
<span>    ColorBySampleDynamicSource <span class="op">=</span> <span class="cn">FALSE</span>, ShapeBy <span class="op">=</span> <span class="st">"None"</span>, SizeBy <span class="op">=</span> <span class="st">"None"</span>, </span>
<span>    SelectionAlpha <span class="op">=</span> <span class="fl">0.1</span>, ZoomData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, BrushData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>    VisualBoxOpen <span class="op">=</span> <span class="cn">FALSE</span>, VisualChoices <span class="op">=</span> <span class="st">"Color"</span>, ContourAdd <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    ContourColor <span class="op">=</span> <span class="st">"#0000FF"</span>, PointSize <span class="op">=</span> <span class="fl">1</span>, PointAlpha <span class="op">=</span> <span class="fl">1</span>, </span>
<span>    Downsample <span class="op">=</span> <span class="cn">FALSE</span>, DownsampleResolution <span class="op">=</span> <span class="fl">200</span>, CustomLabels <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    CustomLabelsText <span class="op">=</span> <span class="st">"SRR2140028"</span>, FontSize <span class="op">=</span> <span class="fl">1</span>, LegendPointSize <span class="op">=</span> <span class="fl">1</span>, </span>
<span>    LegendPosition <span class="op">=</span> <span class="st">"Bottom"</span>, HoverInfo <span class="op">=</span> <span class="cn">TRUE</span>, LabelCenters <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    LabelCentersBy <span class="op">=</span> <span class="st">"driver_1_s"</span>, LabelCentersColor <span class="op">=</span> <span class="st">"#000000"</span>, </span>
<span>    VersionInfo <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>iSEE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/structure.html" class="external-link">structure</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2L</span>, <span class="fl">17L</span>, <span class="fl">1L</span><span class="op">)</span><span class="op">)</span>, class <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"package_version"</span>, </span>
<span>    <span class="st">"numeric_version"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, PanelId <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>ReducedDimensionPlot <span class="op">=</span> <span class="fl">1L</span><span class="op">)</span>, </span>
<span>    PanelHeight <span class="op">=</span> <span class="fl">500L</span>, PanelWidth <span class="op">=</span> <span class="fl">4L</span>, SelectionBoxOpen <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    RowSelectionSource <span class="op">=</span> <span class="st">"---"</span>, ColumnSelectionSource <span class="op">=</span> <span class="st">"---"</span>, </span>
<span>    DataBoxOpen <span class="op">=</span> <span class="cn">FALSE</span>, RowSelectionDynamicSource <span class="op">=</span> <span class="cn">FALSE</span>, ColumnSelectionDynamicSource <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    RowSelectionRestrict <span class="op">=</span> <span class="cn">FALSE</span>, ColumnSelectionRestrict <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    SelectionHistory <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">################################################################################</span></span>
<span><span class="co"># Settings for Feature assay plot 1</span></span>
<span><span class="co">################################################################################</span></span>
<span></span>
<span><span class="va">initial</span><span class="op">[[</span><span class="st">"FeatureAssayPlot1"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">new</span><span class="op">(</span><span class="st">"FeatureAssayPlot"</span>, Assay <span class="op">=</span> <span class="st">"logcounts"</span>, XAxis <span class="op">=</span> <span class="st">"None"</span>, </span>
<span>    XAxisColumnData <span class="op">=</span> <span class="st">"NREADS"</span>, XAxisFeatureName <span class="op">=</span> <span class="st">"0610007P14Rik"</span>, </span>
<span>    XAxisFeatureSource <span class="op">=</span> <span class="st">"---"</span>, XAxisFeatureDynamicSource <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    YAxisFeatureName <span class="op">=</span> <span class="st">"0610007P14Rik"</span>, YAxisFeatureSource <span class="op">=</span> <span class="st">"---"</span>, </span>
<span>    YAxisFeatureDynamicSource <span class="op">=</span> <span class="cn">FALSE</span>, FacetRowByColData <span class="op">=</span> <span class="st">"driver_1_s"</span>, </span>
<span>    FacetColumnByColData <span class="op">=</span> <span class="st">"driver_1_s"</span>, ColorByColumnData <span class="op">=</span> <span class="st">"NREADS"</span>, </span>
<span>    ColorByFeatureNameAssay <span class="op">=</span> <span class="st">"logcounts"</span>, ColorBySampleNameColor <span class="op">=</span> <span class="st">"#FF0000"</span>, </span>
<span>    ShapeByColumnData <span class="op">=</span> <span class="st">"driver_1_s"</span>, SizeByColumnData <span class="op">=</span> <span class="st">"NREADS"</span>, </span>
<span>    TooltipColumnData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, FacetRowBy <span class="op">=</span> <span class="st">"None"</span>, FacetColumnBy <span class="op">=</span> <span class="st">"None"</span>, </span>
<span>    ColorBy <span class="op">=</span> <span class="st">"None"</span>, ColorByDefaultColor <span class="op">=</span> <span class="st">"#000000"</span>, ColorByFeatureName <span class="op">=</span> <span class="st">"0610007P14Rik"</span>, </span>
<span>    ColorByFeatureSource <span class="op">=</span> <span class="st">"---"</span>, ColorByFeatureDynamicSource <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    ColorBySampleName <span class="op">=</span> <span class="st">"SRR2140028"</span>, ColorBySampleSource <span class="op">=</span> <span class="st">"---"</span>, </span>
<span>    ColorBySampleDynamicSource <span class="op">=</span> <span class="cn">FALSE</span>, ShapeBy <span class="op">=</span> <span class="st">"None"</span>, SizeBy <span class="op">=</span> <span class="st">"None"</span>, </span>
<span>    SelectionAlpha <span class="op">=</span> <span class="fl">0.1</span>, ZoomData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, BrushData <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>    VisualBoxOpen <span class="op">=</span> <span class="cn">FALSE</span>, VisualChoices <span class="op">=</span> <span class="st">"Color"</span>, ContourAdd <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    ContourColor <span class="op">=</span> <span class="st">"#0000FF"</span>, PointSize <span class="op">=</span> <span class="fl">1</span>, PointAlpha <span class="op">=</span> <span class="fl">1</span>, </span>
<span>    Downsample <span class="op">=</span> <span class="cn">FALSE</span>, DownsampleResolution <span class="op">=</span> <span class="fl">200</span>, CustomLabels <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    CustomLabelsText <span class="op">=</span> <span class="st">"SRR2140028"</span>, FontSize <span class="op">=</span> <span class="fl">1</span>, LegendPointSize <span class="op">=</span> <span class="fl">1</span>, </span>
<span>    LegendPosition <span class="op">=</span> <span class="st">"Bottom"</span>, HoverInfo <span class="op">=</span> <span class="cn">TRUE</span>, LabelCenters <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    LabelCentersBy <span class="op">=</span> <span class="st">"driver_1_s"</span>, LabelCentersColor <span class="op">=</span> <span class="st">"#000000"</span>, </span>
<span>    VersionInfo <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>iSEE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/structure.html" class="external-link">structure</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2L</span>, <span class="fl">17L</span>, <span class="fl">1L</span><span class="op">)</span><span class="op">)</span>, class <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"package_version"</span>, </span>
<span>    <span class="st">"numeric_version"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, PanelId <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>FeatureAssayPlot <span class="op">=</span> <span class="fl">1L</span><span class="op">)</span>, </span>
<span>    PanelHeight <span class="op">=</span> <span class="fl">500L</span>, PanelWidth <span class="op">=</span> <span class="fl">4L</span>, SelectionBoxOpen <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    RowSelectionSource <span class="op">=</span> <span class="st">"---"</span>, ColumnSelectionSource <span class="op">=</span> <span class="st">"---"</span>, </span>
<span>    DataBoxOpen <span class="op">=</span> <span class="cn">FALSE</span>, RowSelectionDynamicSource <span class="op">=</span> <span class="cn">FALSE</span>, ColumnSelectionDynamicSource <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    RowSelectionRestrict <span class="op">=</span> <span class="cn">FALSE</span>, ColumnSelectionRestrict <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    SelectionHistory <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The above R script creates an object called <code>initial</code> that
contains all the information necessary for the <code><a href="https://rdrr.io/pkg/iSEE/man/iSEE.html" class="external-link">iSEE()</a></code>
function to launch an app in a specific initial state.</p>
<p>All we need to do is:</p>
<ul>
<li>Run that R script, to create the <code>initial</code> object that
represents the initial state of the app</li>
<li>Pass this <code>initial</code> object to the <code>initial=</code>
argument of the <code><a href="https://rdrr.io/pkg/iSEE/man/iSEE.html" class="external-link">iSEE()</a></code> function, to launch a new app in the
desired initial state</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/iSEE/man/iSEE.html" class="external-link">iSEE</a></span><span class="op">(</span><span class="va">sce</span>, initial <span class="op">=</span> <span class="va">initial</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="img/iSEE_initial.png" alt="Screenshot of iSEE app launched in a predefined initial configuration."><div class="figcaption">Screenshot of iSEE app launched in a predefined
initial configuration.</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="existing-isee-extensions">Existing iSEE extensions<a class="anchor" aria-label="anchor" href="#existing-isee-extensions"></a>
</h2>
</div>
<div class="section level2">
<h2 id="overview-2">Overview<a class="anchor" aria-label="anchor" href="#overview-2"></a>
</h2>
<p>In the next 10 minutes, we will:</p>
<ul>
<li>Load and preprocess a new demonstration data set</li>
<li>Launch an iSEE app that includes panels implemented in extension
packages</li>
</ul>
<div class="section level3">
<h3 id="isee-extension-packages">iSEE extension packages<a class="anchor" aria-label="anchor" href="#isee-extension-packages"></a>
</h3>
<table class="table">
<colgroup>
<col width="15%">
<col width="84%">
</colgroup>
<thead><tr class="header">
<th align="left">Package</th>
<th align="left">Purpose</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><a href="https://github.com/iSEE/iSEE" class="external-link">iSEE</a></td>
<td align="left">Core functionality.</td>
</tr>
<tr class="even">
<td align="left"><a href="https://github.com/iSEE/iSEEde" class="external-link">iSEEde</a></td>
<td align="left">Extension for differential expression analysis.</td>
</tr>
<tr class="odd">
<td align="left"><a href="https://github.com/iSEE/iSEEhex" class="external-link">iSEEhex</a></td>
<td align="left">Extension for summarisation into hexagonal bins.</td>
</tr>
<tr class="even">
<td align="left"><a href="https://github.com/iSEE/iSEEhub" class="external-link">iSEEhub</a></td>
<td align="left">Extension for access to the Bioconductor <em><a href="https://bioconductor.org/packages/3.20/ExperimentHub" class="external-link">ExperimentHub</a></em>.</td>
</tr>
<tr class="odd">
<td align="left"><a href="https://github.com/iSEE/iSEEindex" class="external-link">iSEEindex</a></td>
<td align="left">Extension for access to custom collectins of local and
remote data sets.</td>
</tr>
<tr class="even">
<td align="left"><a href="https://github.com/iSEE/iSEEpathways" class="external-link">iSEEpathways</a></td>
<td align="left">Extension for pathway analysis.</td>
</tr>
<tr class="odd">
<td align="left"><a href="https://github.com/iSEE/iSEEu" class="external-link">iSEEu</a></td>
<td align="left">The ‘iSEE universe’: Extension for miscellaneous and
experimental functionality.</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="input-data-1">Input data<a class="anchor" aria-label="anchor" href="#input-data-1"></a>
</h3>
<p>The <em><a href="https://bioconductor.org/packages/3.20/iSEEpathways" class="external-link">iSEEpathways</a></em>
vignette <a href="https://isee.github.io/iSEEpathways/articles/integration.html" class="external-link">Integration
with other panels</a> can also be accessed locally using the R code
<code><a href="https://bioconductor.org/packages/release/bioc/vignettes/iSEEpathways/inst/doc/integration.html" class="external-link">vignette("integration", package = "iSEEpathways")</a></code>.</p>
<p>In this part, we load the <em><a href="https://bioconductor.org/packages/3.20/airway" class="external-link">airway</a></em>
package, providing a <code>RangedSummarizedExperiment</code> object for
RNA-Seq in airway smooth muscle cells.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span><span class="va">airway</span></span>
<span><span class="co">#&gt; class: RangedSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 63677 8 </span></span>
<span><span class="co">#&gt; metadata(1): ''</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(63677): ENSG00000000003 ENSG00000000005 ... ENSG00000273492</span></span>
<span><span class="co">#&gt;   ENSG00000273493</span></span>
<span><span class="co">#&gt; rowData names(10): gene_id gene_name ... seq_coord_system symbol</span></span>
<span><span class="co">#&gt; colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521</span></span>
<span><span class="co">#&gt; colData names(9): SampleName cell ... Sample BioSample</span></span></code></pre></div>
<p>We quickly reorder the levels of the dexamethasone treatment,
ensuring that the untreated level is first, and used as reference level
during the upcoming differential expression analysis.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">airway</span><span class="op">$</span><span class="va">dex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/relevel.html" class="external-link">relevel</a></span><span class="op">(</span><span class="va">airway</span><span class="op">$</span><span class="va">dex</span>, <span class="st">"untrt"</span><span class="op">)</span></span></code></pre></div>
<p>We also take a minute to convert rownames to more recognisable gene
symbols using the annotation package <em><a href="https://bioconductor.org/packages/3.20/org.Hs.eg.db" class="external-link">org.Hs.eg.db</a></em>.</p>
<p>To avoid losing any information, we store a copy of the original
Ensembl gene identifiers and the corresponding gene symbols in the row
metadata.</p>
<p>To make sure that rownames are unique, we use the <em><a href="https://bioconductor.org/packages/3.20/scater" class="external-link">scater</a></em>
function <code><a href="https://rdrr.io/pkg/scuttle/man/uniquifyFeatureNames.html" class="external-link">uniquifyFeatureNames()</a></code>. The function uses the gene
symbol if unique; otherwise it combines it with the Ensembl gene
identifier to make it unique.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"org.Hs.eg.db"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: AnnotationDbi</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://bioconductor.org/packages/scater/" class="external-link">"scater"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">[[</span><span class="st">"ENSEMBL"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">[[</span><span class="st">"SYMBOL"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html" class="external-link">mapIds</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span>, <span class="st">"SYMBOL"</span>, <span class="st">"ENSEMBL"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 'select()' returned 1:many mapping between keys and columns</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">[[</span><span class="st">"uniquifyFeatureNames"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/uniquifyFeatureNames.html" class="external-link">uniquifyFeatureNames</a></span><span class="op">(</span></span>
<span>  ID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">[[</span><span class="st">"ENSEMBL"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>  names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">[[</span><span class="st">"SYMBOL"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">airway</span><span class="op">)</span><span class="op">[[</span><span class="st">"uniquifyFeatureNames"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>Plan:</p>
<ul>
<li>Run code from the iSEEpathways vignette <a href="https://isee.github.io/iSEEpathways/articles/integration.html" class="external-link uri">https://isee.github.io/iSEEpathways/articles/integration.html</a>
</li>
<li>Launch the app</li>
<li>Showcase the integration of pathways, DE, and core panels (e.g.,
select a pathway in the results table, show those genes in the volcano
plot and heat map)</li>
</ul>
<p>(Remember: 10 min)</p>
</div>
</div>
<div class="section level2">
<h2 id="writing-isee-extensions">Writing iSEE extensions<a class="anchor" aria-label="anchor" href="#writing-isee-extensions"></a>
</h2>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p><a href="mailto:kevinrue67@gmail.com" class="email">kevinrue67@gmail.com</a><a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p><a href="mailto:marinif@uni-mainz.de" class="email">marinif@uni-mainz.de</a><a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p><a href="mailto:charlottesoneson@gmail.com" class="email">charlottesoneson@gmail.com</a><a href="#fnref3" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Rue-Albrecht, Federico Marini, Charlotte Soneson.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
